core/src/test/java/org/elasticsearch/cluster/routing/DelayedAllocationServiceTests.java

```diff
@@ -0,0 +1,473 @@
+import static org.elasticsearch.cluster.routing.ShardRoutingState.STARTED;
+import static org.elasticsearch.common.unit.TimeValue.timeValueMillis;
+import static org.elasticsearch.common.unit.TimeValue.timeValueSeconds;
+import static org.hamcrest.Matchers.equalTo;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+
+/**
+ */
+public class DelayedAllocationServiceTests extends ESAllocationTestCase {
+
+    private TestDelayAllocationService delayedAllocationService;
+    private MockAllocationService allocationService;
+    private ClusterService clusterService;
+    private ThreadPool threadPool;
+
+    @Before
+    public void createDelayedAllocationService() {
+        threadPool = new ThreadPool(getTestName());
+        clusterService = mock(ClusterService.class);
+        allocationService = createAllocationService(Settings.EMPTY, new DelayedShardsMockGatewayAllocator());
+        delayedAllocationService = new TestDelayAllocationService(Settings.EMPTY, threadPool, clusterService, allocationService);
+        verify(clusterService).addFirst(delayedAllocationService);
+    }
+
+    @After
+    public void shutdownThreadPool() throws Exception {
+        terminate(threadPool);
+    }
+
+    public void testNoDelayedUnassigned() throws Exception {
+        MetaData metaData = MetaData.builder()
+            .put(IndexMetaData.builder("test").settings(settings(Version.CURRENT)
+                .put(UnassignedInfo.INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.getKey(), "0"))
+                .numberOfShards(1).numberOfReplicas(1))
+            .build();
+        ClusterState clusterState = ClusterState.builder(ClusterName.DEFAULT)
+            .metaData(metaData)
+            .routingTable(RoutingTable.builder().addAsNew(metaData.index("test")).build()).build();
+        clusterState = ClusterState.builder(clusterState)
+            .nodes(DiscoveryNodes.builder().put(newNode("node1")).put(newNode("node2")).localNodeId("node1").masterNodeId("node1"))
+            .build();
+        clusterState = ClusterState.builder(clusterState).routingResult(allocationService.reroute(clusterState, "reroute")).build();
+        // starting primaries
+        clusterState = ClusterState.builder(clusterState)
+            .routingResult(allocationService.applyStartedShards(clusterState, clusterState.getRoutingNodes().shardsWithState(INITIALIZING)))
+            .build();
+        // starting replicas
+        clusterState = ClusterState.builder(clusterState)
+            .routingResult(allocationService.applyStartedShards(clusterState, clusterState.getRoutingNodes().shardsWithState(INITIALIZING)))
+            .build();
+        assertThat(clusterState.getRoutingNodes().unassigned().size() > 0, equalTo(false));
+        // remove node2 and reroute
```

Comment:
I'm wondering if we want to randomly add a third node and see that we still allocate (not only say "delayed" is false) - maybe this is covered by another test. If so , no need for this here